### ISSUE TITLE
TAMUNA

### ISSUE THREAD

**Issue by jafermarq** [adap/flower#2033]
# TAMUNA

* **Title:** TAMUNA: Doubly Accelerated Federated Learning with Local Training, Compression, and Partial Participation
* **Venue:** -
* **Link to paper:** https://arxiv.org/abs/2302.09832

## Do you want to work on this baseline? 

> 🌻 Check everything about the `Summer of Reproducibility` on [`flower.dev/summer`](https://flower.dev/summer/)

> All available baselines are listed in the [Summer of Reproducibility Dashboard](https://github.com/orgs/adap/projects/6/) and also in the [GitHub Issues with the `summer-of-reproducibility` label](https://github.com/adap/flower/labels/summer-of-reproducibility). The content is the same.

> 📝 It is advised to complete these steps before your start working on your code. But if you can't wait to implement your baseline with Flower (we totally understand it 😄), please ensure you follow the steps on [how to contribute a new baseline](https://github.com/adap/flower/tree/main/baselines#contributing-a-new-baseline).

What follows are the steps 1 & 2 in the Summer of Reproducibility instructions.

### 1. Join the Summer of Reproducibility program
- [x] Join the [Flower Slack](https://flower.dev/join-slack/) and say _"hi! 👋"_ in the channel `#summer-of-reproducibility`.
- [x] Pick a baseline from our curated list **<---------------------------------------- [you are doing this now]**
### 2. Define the scope of your contribution
- [x] **What are you going to reproduce?** Add a comment to your issue and tell us about your plan regarding this baseline: what experiments from the paper are you reproducing?, for which datasets ? the more details you provide us with the better !
- [x] **Check if you are eligible for a reward**. 
    > As we have to comply with US/EU regulations, we have checked that individual contributors based on these countries or territories are eligible: Australia, Austria, Belgium, Bulgaria, Canada, Croatia, Cyprus, Czech Republic, Denmark, Estonia, Finland, France, Germany, Gibraltar, Greece, Hong Kong SAR China, Hungary, India, Ireland, Italy, Japan, Latvia, Liechtenstein, Lithuania, Luxembourg, Malta, Mexico, Netherlands, New Zealand, Norway, Poland, Portugal, Romania, Singapore, Slovakia, Slovenia, South Korea, Spain, Sweden, Switzerland, Thailand, United Arab emirates, United Kingdom United States.

    If where you are based is not on the list, please send us an email (`summer@flower.dev`) letting us know a bit about yourself (where are you currently based?, are you a university student? do you work at a public institution?). Please tell us the baselines you are interested in implementing (i.e. tell us your GitHub issue if you have crated one). We will reach back to you.
- [x] We will discuss with you about your contribution plan, if it sounds like a substantial enough contribution according to the Summer of Reproducibility rules (check our website [flower.dev/summer](https://flower.dev/summer/)), you'll get the OK from us to start working on your baseline!

## What happens next?

- [x] This item will be moved to the `In Progress` stage by a member of the Flower Team.

- [ ] Follow the [instructions for creating a new baseline](https://github.com/adap/flower/tree/main/baselines#contributing-a-new-baseline) which will guide you through the process step-by-step.

## Is something wrong or not clear ?
* Ask a question directly in your issue.
* Reach out to us via the [Flower Slack](https://flower.dev/join-slack/) and ask your question in the #summer-of-reproducibility channel
* Check all the details (including FAQ) in the Summer of Reproducibility website: flower.dev/summer

**Comment by Crabzmatic** [comment]
Hi! I'm Ivan Agarský and together with Grigory Malinovsky @gsmalinovsky we want to reproduce the results from the paper _TAMUNA: Doubly Accelerated Federated Learning with Local Training, Compression, and Partial Participation_ (https://arxiv.org/abs/2302.09832). More specifically, Figure 2 and Figure 3 from the paper.

Optionally, we may want to also do some DNN experiments.

**Comment by Crabzmatic** [comment]
The paper contains Logistic Regression experiments on datasets from LIBSVM. It does not use PyTorch or any similar frameworks besides NumPy. If we add it to flower in this form, it won't be possible to use it as baseline for NN experiments. Is this okay, or should all baselines be implemented with NN in mind?

**Comment by jafermarq** [comment]
hi @Crabzmatic , it's great to se you and Grigory want to implement `TAMUNA`. Figure 2 and 3 make sense to me. All the baselines do contain NNs (some deeper than others) with the exception of `scaffnew` if I remember correctly now. If porting `TAMUNA` to NNs is not feasible (I presume it is possible but not with major effort?), how about designing an experiment around a dataset that would likely be familiar to others interested also in comms compression for FL? for instance MNIST, which is also part of `LIBSVM`?

**Comment by jafermarq** [comment]
cool! I have now ✅ all the points above for Step 1 & 2 and adde you both as _assignees_ to this issue. You can find how to start with the code for your baseline following the link in the _What happens next?_ section above. Feel free to reach out to me over Slack if you have some questions!

Looking forward to seeing `TAMUNA` in action ! 🚀

**Comment by Crabzmatic** [comment]
Hi, we will need to use stateful clients for `TAMUNA`. I noticed there are two open PRs (https://github.com/adap/flower/pull/2148, https://github.com/adap/flower/pull/2160) that add stateful clients support. Do you have a time frame for finalizing these PRs?

**Comment by jafermarq** [comment]
Hi @Crabzmatic , yes we are aware a couple of baselines in the _Summer of Reproducibility_ use stateful clients. Our aim is to finalise these baselines in the coming 1-2 weeks to provide native support. In the meantime I'd encourage you to implement your own mechanism to keep the state on the client. For instance, if you run your simulation on a single node, an easy way of achieving so is to construct a directory where to keep the states (at the beginning of the experiment), then each client upon initialisation fetches it's own state (maybe stored in the file system as a pickle with a unique identifier in the name of the file, e.g. the client id `cid`). When client updates its state, it can be written back to the file. This is probably the easiest way of achieving stateful clients at the moment.

**Comment by jafermarq** [comment]
Congratulations @Crabzmatic @gsmalinovsky ! your baseline has been merged. We'll follow up with you soon to complete your Summer of Reproducibility journey 😄 !
